signac/core/synced_list.py

```diff
@@ -0,0 +1,172 @@
+# Copyright (c) 2020 The Regents of the University of Michigan
+# All rights reserved.
+# This software is licensed under the BSD 3-Clause License.
+"""SyncedList."""
+
+from collections.abc import Sequence
+from collections.abc import MutableSequence
+
+from .synced_collection import SyncedCollection
+from .synced_collection import NUMPY
+
+if NUMPY:
+    import numpy
+
+
+class SyncedList(SyncedCollection, MutableSequence):
+    """Implementation of list data structure.
+
+    The SyncedList inherits from :class:`~core.collection_api.SyncedCollection`
+    and :class:`~collections.abc.MutableSequence`. Therefore, it behaves similar
+    to a :class:`list`.
+
+    .. warning::
+
+        While the SyncedList object behaves like a dictionary, there are
+        important distinctions to remember. In particular, because operations
+        are reflected as changes to an underlying file, copying (even deep
+        copying) a SyncedList instance may exhibit unexpected behavior. If a
+        true copy is required, you should use the `to_base()` method to get a
+        dictionary representation, and if necessary construct a new SyncedList.
+    """
+
+    def __init__(self, data=None, **kwargs):
+        super().__init__(**kwargs)
+        if data is None:
+            self._data = []
+        else:
+            if NUMPY and isinstance(data, numpy.ndarray):
+                data = data.tolist()
+            with self._suspend_sync():
+                self._data = [self.from_base(data=value, parent=self) for value in data]
+            self.sync()
+
+    @classmethod
+    def is_base_type(cls, data):
+        """Check whether the data is an non-string Sequence.
+
+        Parameters
+        ----------
+        data: any
+            Data to be checked
+
+        Returns
+        -------
+        bool
+        """
+        if isinstance(data, Sequence) and not isinstance(data, str):
+            return True
+        elif NUMPY:
+            if isinstance(data, numpy.ndarray):
+                return True
+        return False
+
+    def to_base(self):
+        """Convert the SyncedList object to list.
+
+        Returns
+        -------
+        converted: list
+            List containing the conveted SyncedList object.
+        """
+        converted = list()
+        for value in self._data:
+            if isinstance(value, SyncedCollection):
+                converted.append(value.to_base())
+            else:
+                converted.append(value)
+        return converted
+
+    def _update(self, data=None):
+        """Update the instance of SyncedList with data by using dfs."""
+        if data is None:
+            data = []
+        if isinstance(data, Sequence) and not isinstance(data, str):
+            with self._suspend_sync():
+                for i in range(min(len(self), len(data))):
```

Comment:
Please include a description of the algorithm that is used here in a code comment. I don't know if my proposed comment is correct, but marking it with `TODO:` could also help if there's room for future improvement.
```suggestion
                # This loop avoids rebuilding existing synced collections for performance.
                # TODO: Potential improvements to this code are...
                for i in range(min(len(self), len(data))):
```
Previous discussion: https://github.com/glotzerlab/signac/pull/336#discussion_r446540980